### PR TITLE
Make TypeScript types conforms to strict mode

### PR DESCRIPTION
**Pull request by pmmmwh**
This PR tries to make all code within the `source` folder conforms to strict mode.

There are some parts that I had to trace the source code and make educated guesses as I am uncertain with the actual available values, so please do tell me if I am wrong! Also - there are some places we could either use an intermediate variable or use non-null assertions. Please tell me which one you would prefer to maintain consistency 👀

#### Next Steps
- [x] Confirm source code is running fine
- [x] Fix types in tests
- [ ] Fix ts-ignore and typescript eslint errors
- [ ] Check for potential `const` assertions that can be applied

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [ ] I have included some tests.
- [ ] If it's a new feature, I have included documentation updates.

### CODE REVIEW

**Comment by szmarczak** [comment]
Beware of #921 - if it gets merged you'll have to rework many many things :P I'd prefer if you waited for its merge.

**Comment by pmmmwh** [comment]
> Beware of #921 - if it gets merged you'll have to rework many many things :P I'd prefer if you waited for its merge.

Hmm. Actually I have kinda finished the whole process of typing stuff. I think maybe I'll push first, then wait for its merge, and then tackle the remaining parts that are broken after that? I wouldn't expect a review to be coming soon, but maybe some of this code can be referenced in that PR too?

**Comment by szmarczak** [comment]
TypeScript needs to ignore files in `tests/*` because they are not TypeScript yet.

File: source/as-promise.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ type ResponseReturn = Response | Buffer | string | any;
 
 export const isProxiedSymbol: unique symbol = Symbol('proxied');
 
-export default function asPromise(options: NormalizedOptions): CancelableRequest<Response> {
+export default function asPromise(options: NormalizedOptions): CancelableRequest<Response> | Response['body'] {
```
Comment (szmarczak) [inline source/as-promise.ts]: Good catch! :)

File: source/as-promise.ts [hunk 2]
```diff
@@ -56,7 +56,7 @@ export default function asPromise(options: NormalizedOptions): CancelableRequest
 				return;
 			}
 
-			if (response.req?.aborted) {
+			if (response.req.aborted) {
```
Comment (szmarczak) [inline source/as-promise.ts]:
```suggestion
			if (response.req?.aborted) {
```

Because the cached response doesn't have `req` property

File: source/as-promise.ts [hunk 3]
```diff
@@ -107,7 +107,7 @@ export default function asPromise(options: NormalizedOptions): CancelableRequest
 
 			if (statusCode !== 304 && (statusCode < 200 || statusCode > limitStatusCode)) {
 				const error = new HTTPError(response, options);
-				if (emitter.retry(error) === false) {
+				if (!emitter.retry(error)) {
```
Comment (szmarczak) [inline source/as-promise.ts]: :+1:

File: source/calculate-retry-delay.ts [hunk 4]
```diff
@@ -1,40 +1,47 @@
...
-	if (response && Reflect.has(response.headers, 'retry-after') && retryAfterStatusCodes.has(response.statusCode)) {
-		let after = Number(response.headers['retry-after']);
-		if (is.nan(after)) {
-			after = Date.parse(response.headers['retry-after']) - Date.now();
-		} else {
-			after *= 1000;
-		}
-
-		if (after > retryOptions.maxRetryAfter) {
+	if (error instanceof GotError) {
```
Comment (szmarczak) [inline source/calculate-retry-delay.ts]: Isn't this always true?

File: source/create.ts [hunk 5]
```diff
@@ -26,15 +27,17 @@ export type HTTPAlias =
 	| 'head'
 	| 'delete';
 
-export type ReturnResponse = (url: URLArgument | Options & {stream?: false; url: URLArgument}, options?: Options & {stream?: false}) => CancelableRequest<Response>;
-export type ReturnStream = (url: URLArgument | Options & {stream: true; url: URLArgument}, options?: Options & {stream: true}) => ProxyStream;
-export type GotReturn = ProxyStream | CancelableRequest<Response>;
+export type ReturnResponse = (url: URLOrOptions | Options & {stream?: false}, options?: Options & {stream?: false}) => ReturnType<typeof asPromise>;
+export type ReturnStream = (url: URLOrOptions | Options & {stream?: true}, options?: Options & {stream?: true}) => ReturnType<typeof asStream>;
+export type GotReturn = ReturnType<ReturnResponse> | ReturnType<ReturnStream>;
```
Comment (szmarczak) [inline source/create.ts]: :+1:

File: source/create.ts [hunk 6]
```diff
@@ -131,24 +144,24 @@ const create = (nonNormalizedDefaults: Defaults): Got => {
...
+		const options: Array<PartialDeep<DefaultOptions> & Options> = [defaults.options];
 		const handlers: HandlerFunction[] = [...defaults.handlers];
-		let mutableDefaults: boolean;
+		let mutableDefaults: boolean | undefined;
 
 		for (const value of instancesOrOptions) {
-			if (Reflect.has(value, 'defaults')) {
-				options.push((value as Got).defaults.options);
-				handlers.push(...(value as Got).defaults.handlers.filter(handler => handler !== defaultHandler));
+			if (isGotInstance(value)) {
```
Comment (szmarczak) [inline source/create.ts]: :+1: nice work here

File: source/utils/timed-out.ts [hunk 7]
```diff
@@ -85,11 +85,12 @@ export default (request: ClientRequest, delays: Delays, options: TimedOutOptions
 	}
 
 	if (typeof delays.socket !== 'undefined') {
+		const {socket} = delays;
```
Comment (szmarczak) [inline source/utils/timed-out.ts]: No need to extract `socket` for only one use. Same below.

File: source/calculate-retry-delay.ts (hunk 4 above)
Comment (pmmmwh) [reply source/calculate-retry-delay.ts]: I am not sure - in tests I see a check on user-defined custom error codes, so it might not be?

File: source/as-promise.ts (hunk 2 above)
Comment (pmmmwh) [reply source/as-promise.ts]: Okies - will reflect this change in the `Response` type too!

File: source/create.ts (hunk 5 above)
Comment (pmmmwh) [reply source/create.ts]: Actually, this can be further slimmed down in the instance type for `Got` below. One could use `Parameters<T>` and `ResponseType<T>`, but I am not sure if it will hurt readability in the users' side.

File: source/utils/timed-out.ts (hunk 7 above)
Comment (pmmmwh) [reply source/utils/timed-out.ts]: The reason I did this change is that both TypeScript (and me) is confused on whether issues due to time will occur, since the code below is defined within a callback. If the object will be always defined, maybe I can switch to using non-null assertions instead.

**Comment by pmmmwh** [comment]
> TypeScript needs to ignore files in `tests/*` because they are not TypeScript yet.

Actually after the changes in the source types tests aren't too difficult to type anymore, so I also typed them. I verified by running `tsc --noEmit --pretty` against the whole project and no errors are reported, so I am not sure if I am confusing `ts-node` by doing nothing, because the project compiles ...

File: source/calculate-retry-delay.ts (hunk 4 above)
Comment (szmarczak) [reply source/calculate-retry-delay.ts]: But it's still GotError, isn't it? It's created in `onError` function IIRC.

File: source/as-promise.ts (hunk 2 above)
Comment (pmmmwh) [reply source/as-promise.ts]: Done :)

File: source/calculate-retry-delay.ts (hunk 4 above)
Comment (pmmmwh) [reply source/calculate-retry-delay.ts]: I checked the source code - it seems that there is only one place calling `calculateRetryDelay`: `emitter.retry`. It is being called in two places, with the possible error choices `TimeoutError`, `HTTPError` and `RequestError`. I think you are right - I will update to reflect this in the code.

File: source/utils/timed-out.ts (hunk 7 above)
Comment (pmmmwh) [reply source/utils/timed-out.ts]: Fixed.

**Comment by pmmmwh** [comment]
So I fixed the issue - apparently `ts-node` will need to use `typeRoots` to correctly read type definitions, so I also did that move in the PR. Now I wonder why tests are failing on CI for node 10 and 12, because they apparently passes on my local machine 🤔

File: source/utils/types.ts [hunk 8]
```diff
@@ -113,10 +119,27 @@ export interface Delays {
 	request?: number;
 }
 
-export type Headers = Record<string, string | string[]>;
+export type Headers = Record<string, string | string[] | null | undefined>;
```
Comment (sindresorhus) [inline source/utils/types.ts]:
```suggestion
export type Headers = Record<string, string | string[] | undefined>;
```

We decided to only allow `undefined` to clear headers and not `null`.

File: test/cookies.ts [hunk 9]
```diff
@@ -131,11 +130,11 @@ test('no unhandled errors', async t => {
 	const options = {
 		cookieJar: {
 			setCookie: () => {},
-			getCookieString: (_, cb) => cb(new Error(message))
-		}
+			getCookieString: (_: string, cb: (err: Error) => void) => cb(new Error(message))
```
Comment (sindresorhus) [inline test/cookies.ts]:
```suggestion
			getCookieString: (_: string, callback: (error: Error) => void) => callback(new Error(message))
```

File: package.json [hunk 10]
```diff
@@ -39,8 +39,6 @@
 	"dependencies": {
 		"@sindresorhus/is": "^1.0.0",
 		"@szmarczak/http-timer": "^3.0.0",
-		"@types/cacheable-request": "^6.0.1",
-		"@types/tough-cookie": "^2.3.5",
```
Comment (sindresorhus) [inline package.json]: I think these needs to be dependencies as they're used in the actual code.

Comment (szmarczak) [reply package.json]: https://stackoverflow.com/a/45176764

Comment (szmarczak) [reply package.json]: Only TypeScript users need this.

File: source/errors.ts [hunk 11]
```diff
@@ -45,41 +46,42 @@ export class ReadError extends GotError {
...
-	readonly response: Response;
+	readonly response!: Response;
 
 	constructor(error: Error, response: Response, options: NormalizedOptions) {
-		super(`${error.message} in "${format(options as unknown as URL)}"`, error, options);
+		super(`${error.message} in "${format(options as unknown as URL | UrlObject)}"`, error, options);
 		this.name = 'ParseError';
 
 		Object.defineProperty(this, 'response', {
+			enumerable: false,
```
[the remark this answers is not in the document]
Comment (pmmmwh) [reply source/errors.ts]: Since TS 3.7 with the `useDefineForClassFields` flag on, all class defined properties will be initialized with `enumerable: true`, `configurable: true`, and `writable: true`. I didn't realize I could mitigate this with the declare handler to mitigate that, will fix that in a commit.

File: source/known-hook-events.ts [hunk 12]
```diff
@@ -102,6 +102,7 @@ export interface Hooks {
 	@default []
 	*/
 	afterResponse?: AfterResponseHook[];
+	[hookName: string]: any;
```
[the remark this answers is not in the document]
Comment (pmmmwh) [reply source/known-hook-events.ts]: I added this after seeing a test on "allows extra keys in `options.hooks`".

Is the correct behavior as follows?
1. Allow custom keys passed into hooks
2. Normalized/merged options only contains known hook events

File: source/errors.ts [hunk 13]
```diff
@@ -45,41 +46,42 @@ export class ReadError extends GotError {
 }
 
 export class ParseError extends GotError {
-	readonly response: Response;
+	readonly response!: Response;
```
[the remark this answers is not in the document]
Comment (pmmmwh) [reply source/errors.ts]: Typescript complains that response is not initialized and definitely assigned in the constructor, because it won't pickup `Object.defineProperty` calls as a property assignment. I added this since from the source it is known that this will be defined within the constructor.

File: source/utils/types.ts (hunk 8 above)
Comment (pmmmwh) [reply source/utils/types.ts]: This will need to update the test on removing null headers - should I also do it?

File: package.json (hunk 10 above)
Comment (pmmmwh) [reply package.json]: > Only TypeScript users need this.

Do you think I should include them in dependencies since our exported type definitions includes them?

File: source/errors.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (pmmmwh) [reply source/errors.ts]: OK ... so I realized why I didn't do that earlier. Apparently this feature is [broken](https://github.com/microsoft/TypeScript/issues/34972). Should we wait on that, or just add a comment to ignore that for now and move on?

File: test/cookies.ts (hunk 9 above)
Comment (pmmmwh) [reply test/cookies.ts]: Fixed!

**Comment by pmmmwh** [comment]
I am currently in the process of fixing merge conflicts. There's one thing I would like to ask though - can you enlighten me on why the global `URL` and `URLSearchParams` objects are preferred instead of the ones from the `url` module? I am just asking because I am curious, and this does generate quite a few cases where URL !== URL and URLSearchParams !== URLSearchParams.

**Comment by szmarczak** [comment]
> URL !== URL and URLSearchParams !== URLSearchParams

`// @ts-ignore` that. It's a well-known bug in TypeScript. 

> why the global URL and URLSearchParams objects are preferred instead of the ones from the url module?

Because `require('url')` is not necessary to get URL and URLSearchParams. It's to prevent moot code.

File: package.json (hunk 10 above)
Comment (szmarczak) [reply package.json]: I think that we shouldn't include the types as dependencies. They should be in devDependencies, because people that don't use TS will have to download 1.5MB+ data. People that want to use TS would need to `npm install @types/... @types/... @types/...`.

File: source/errors.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/errors.ts]: > Should we wait on that, or just add a comment to ignore that for now and move on?

Just add a comment that links to https://github.com/microsoft/TypeScript/issues/34972 and move on.

File: source/errors.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/errors.ts]: Alright. Thanks for explaining.

**Comment by sindresorhus** [comment]
> Because require('url') is not necessary to get URL and URLSearchParams. It's to prevent moot code.

Explicitly importing them is only a legacy thing. But the main reason is consistency with browsers where they are only globals.

File: package.json (hunk 10 above)
Comment (sindresorhus) [reply package.json]: Actually, I was not entirely correct. `@types/tough-cookie` can be a dev dependency as we don't directly depend on it, and it's up to the user to add it if they use `tough-cookie`. However, we do depend on `cacheable-request`, and we use its types, so I think we need to have `@types/cacheable-request` as a dependency, otherwise Got will not run when published and used in a TS project, right? Unless I'm missing something.

File: source/utils/types.ts (hunk 8 above)
Comment (sindresorhus) [reply source/utils/types.ts]: Yes, thanks :)

File: source/known-hook-events.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/known-hook-events.ts]: It was added in https://github.com/sindresorhus/got/commit/107756fc28dce579c521b15248f8bba730ae9e54. I think it was just a mistake. We definitely don't want to allow random keys. That makes no sense to me. @szmarczak ?

File: package.json (hunk 10 above)
Comment (szmarczak) [reply package.json]: Many people don't use TS and for them it's only bloat. NPM should somehow detect typescript projects and install types automatically...

File: source/known-hook-events.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (szmarczak) [reply source/known-hook-events.ts]: I agree with @sindresorhus since we have a `context` option for this :) the test need to be done removed then.

File: package.json (hunk 10 above)
Comment (pmmmwh) [reply package.json]: > Actually, I was not entirely correct. `@types/tough-cookie` can be a dev dependency as we don't directly depend on it, and it's up to the user to add it if they use `tough-cookie`. However, we do depend on `cacheable-request`, and we use its types, so I think we need to have `@types/cacheable-request` as a dependency, otherwise Got will not run when published and used in a TS project, right? Unless I'm missing something.

I am going to do this. I think both of you are correct, but maintaining out-of-the-box working status is important for TS users too, so the bloat might be inevitable ...

File: source/as-promise.ts [hunk 14]
```diff
@@ -28,16 +27,16 @@ const parseBody = (body: Response['body'], responseType: NormalizedOptions['resp
 	throw new Error(`Failed to parse body of type '${typeof body}' as '${responseType}'`);
 };
 
-export default function asPromise(options: NormalizedOptions) {
+export default function asPromise<T extends any>(options: NormalizedOptions): CancelableRequest<T> {
```
Comment (sindresorhus) [inline source/as-promise.ts]: `any` => `unknown`, if possible

File: source/known-hook-events.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (pmmmwh) [reply source/known-hook-events.ts]: I have removed the key and added a comment on the test saying that we won't throw but the hook won't get passed in.

File: test/create.ts [hunk 15]
```diff
@@ -27,7 +32,7 @@ test('supports instance defaults', withServer, async (t, server, got) => {
 			'user-agent': 'custom-ua-string'
 		}
 	});
-	const headers = await instance('').json();
+	const headers = await instance('').json() as Headers;
```
Comment (szmarczak) [inline test/create.ts]:
```suggestion
	const headers = await instance('').json<Headers>();
```

**Comment by szmarczak** [comment]
I start to hate TypeScript.

File: test/create.ts (hunk 15 above)
Comment (pmmmwh) [reply test/create.ts]: This couldn't be done because TS complains `Headers` is not assignable to `JsonValue`. I am still thinking of a way to solve that ...

I tried both ways below with no luck:
```js
const headers = await instance('').json<Headers>();
```

```js
const headers: Headers = await instance('').json();
```

**Comment by pmmmwh** [comment]
Okay, with a lot of try and errors I made tests run on Node 10/12 on Travis.
One test failed. I checked the history and I'm pretty sure that I didn't touch that part of the code.

I used `JsonValue` to type JSON responses, but it seems to have some side-effect and will complain any object like interface !== JsonValue. I'll probably think about that tomorrow.

**Comment by pmmmwh** [comment]
> I start to hate TypeScript.

I saw @sindresorhus tweet about it. TS is still short when to infer/track types when we are in a highly-mutable, very functional (with recursions) world, or when there are a lot of overloads to the same function. It does offer a lot of promise, but when a lot of stuff is unknown (especially true for request body), it is so tempting to just go with `any`.

File: source/as-promise.ts [hunk 16]
```diff
@@ -27,7 +27,7 @@ const parseBody = (body: Response['body'], responseType: NormalizedOptions['resp
 	throw new Error(`Failed to parse body of type '${typeof body}' as '${responseType}'`);
 };
 
-export default function asPromise<T extends any>(options: NormalizedOptions): CancelableRequest<T> {
+export default function asPromise<T extends unknown = unknown>(options: NormalizedOptions): CancelableRequest<T> {
```
Comment (sindresorhus) [inline source/as-promise.ts]: I don't think you need to constrain these to `unknown`. The default should be enough: https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/#breaking-changes I think it could be just `<T>(..`. Same in the other places we have this.

File: source/create.ts [hunk 17]
```diff
@@ -23,38 +23,37 @@ export type HTTPAlias =
...
 type OptionsOfBufferResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'buffer'};
 type ResponseBodyOnly = {resolveBodyOnly: true};
 
 interface GotFunctions {
 	// `asPromise` usage
-	(url: string | OptionsOfDefaultResponseBody, options?: OptionsOfDefaultResponseBody): CancelableRequest<Response>;
+	<T extends unknown = unknown>(url: string | OptionsOfDefaultResponseBody, options?: OptionsOfDefaultResponseBody): CancelableRequest<Response<T>>;
 	(url: string | OptionsOfTextResponseBody, options?: OptionsOfTextResponseBody): CancelableRequest<Response<string>>;
-	(url: string | OptionsOfJSONResponseBody, options?: OptionsOfJSONResponseBody): CancelableRequest<Response<object>>;
+	<T extends JsonValue = JsonValue>(url: string | OptionsOfJSONResponseBody, options?: OptionsOfJSONResponseBody): CancelableRequest<Response<T>>;
```
Comment (sindresorhus) [inline source/create.ts]: Let's not constrain this. `JsonValue` type creates more problems than it solves. See: https://github.com/sindresorhus/ky/blob/2f37c3f999efb36db9108893b8b3d4b3a7f5ec45/index.d.ts#L303-L326 (and related links) Applies to other places `JsonValue` is used too.

File: source/utils/types.ts [hunk 18]
```diff
@@ -117,7 +117,7 @@ export interface Delays {
 	request?: number;
 }
 
-export type Headers = Record<string, string | string[] | undefined>;
+export type Headers = Record<string, string | string[] | number | undefined>;
```
Comment (sindresorhus) [inline source/utils/types.ts]: I don't think we should support `number` here. Headers are strings, so I think it's better to force users to explicitly coerce/cast to a string themselves.

File: source/utils/types.ts [hunk 19]
```diff
@@ -173,7 +174,7 @@ export interface Options extends PartialDeep<DefaultOptions>, URLOptions, Except
 	useElectronNet?: boolean;
 	form?: Record<string, any>;
 	json?: Record<string, any>;
-	context?: {[key: string]: unknown};
+	context?: Record<string, any>;
```
Comment (sindresorhus) [inline source/utils/types.ts]: `{[key: string]: any}`

File: test/cache.ts [hunk 20]
```diff
@@ -44,7 +44,7 @@ test('cached response is re-encoded to current encoding option', withServer, asy
 	const firstEncoding = 'base64';
 	const secondEncoding = 'hex';
 
-	const firstResponse = await got({cache, encoding: firstEncoding});
+	const firstResponse = await got<string>({cache, encoding: firstEncoding});
```
Comment (sindresorhus) [inline test/cache.ts]: `got` by default returns a string, so I think it should automatically infer a `string` here. It should also have overloads to change the type depending on the `responseType` option.

File: test/create.ts (hunk 15 above)
Comment (sindresorhus) [reply test/create.ts]: Like I've commented elsewhere, don't use `JsonValue`. It creates more problems than it solves.

Generally, don't use `as` unless absolutely needed, as it can hide problem, since it just force casts anything to anything.

File: test/create.ts [hunk 21]
```diff
@@ -228,21 +228,21 @@ test('extend with custom handlers', withServer, async (t, server, got) => {
...
+	t.is(b.defaults.options.prefixUrl!.toString(), 'https://example.com/');
 });
 
 test('extend with a chain', t => {
 	const a = got.extend({prefixUrl: 'https://example.com/'});
 	const b = got.extend(a, {headers: {foo: 'bar'}});
-	t.is(b.defaults.options.prefixUrl.toString(), 'https://example.com/');
-	t.is(b.defaults.options.headers.foo, 'bar');
+	t.is(b.defaults.options.prefixUrl!.toString(), 'https://example.com/');
+	t.is(b.defaults.options.headers!.foo, 'bar');
```
Comment (sindresorhus) [inline test/create.ts]: `.headers` should always be defined. The default should just be an empty object.

File: test/hooks.ts [hunk 22]
```diff
@@ -258,7 +258,7 @@ test('beforeRedirect is called with options and response', withServer, async (t,
 					t.is(options.url.hostname, 'localhost');
 
 					t.is(response.statusCode, 302);
-					t.is(new URL(response.url).pathname, '/redirect');
+					t.is(new URL(response.url!).pathname, '/redirect');
```
Comment (sindresorhus) [inline test/hooks.ts]: `url` is not documented to possibly be `undefined`: https://github.com/sindresorhus/got#url-1

File: test/progress.ts [hunk 23]
```diff
@@ -72,7 +72,7 @@ test('download progress', withServer, async (t, server, got) => {
 
 	const events: Progress[] = [];
 
-	const {body} = await got({encoding: null})
+	const {body} = await got<Buffer>({encoding: null})
```
Comment (sindresorhus) [inline test/progress.ts]: Note: https://github.com/sindresorhus/got/commit/6cc3d9f3027f28ce4dcbe8605d104e65f7e66787

File: source/request-as-event-emitter.ts [hunk 24]
```diff
@@ -12,7 +12,7 @@ import getResponse from './get-response';
 import {normalizeRequestArguments} from './normalize-arguments';
 import {createProgressStream} from './progress';
 import timedOut, {TimeoutError as TimedOutTimeoutError} from './utils/timed-out';
-import {NormalizedOptions, Response, ResponseObject} from './utils/types';
+import {GenericError, NormalizedOptions, Response, ResponseObject} from './utils/types';
```
Comment (sindresorhus) [inline source/request-as-event-emitter.ts]: I don't like the name `GenericError` as "generic" means something different with types. Maybe `BaseError` or `GeneralError`?

File: source/utils/types.ts (hunk 19 above)
Comment (pmmmwh) [reply source/utils/types.ts]: Do we need to change form/json too?

Comment (sindresorhus) [reply source/utils/types.ts]: Yes. `{[key: string]: any}` has special behavior in TypeScript. https://github.com/sindresorhus/typescript-definition-style-guide/pull/8/files `Record` is fine as a input value where the value is not `any`/`unknown` or a return value. For this reason, I just prefer the `{[key: string]: any}` syntax as I can just consistently use it. Yet another thing I dislike about TypeScript. So much inconsistent and hidden behavior.

File: source/utils/types.ts (hunk 18 above)
Comment (pmmmwh) [reply source/utils/types.ts]: This is referenced from the test for `content-length: 0`. Should I alter the test?

File: source/request-as-event-emitter.ts (hunk 24 above)
Comment (pmmmwh) [reply source/request-as-event-emitter.ts]: Renamed to `GeneralError`.

File: test/progress.ts (hunk 23 above)
Comment (pmmmwh) [reply test/progress.ts]: Merged and fixed 🤞

File: test/hooks.ts (hunk 22 above)
Comment (pmmmwh) [reply test/hooks.ts]: Fixed.

File: test/create.ts (hunk 21 above)
Comment (pmmmwh) [reply test/create.ts]: Fixed!

File: test/cache.ts (hunk 20 above)
Comment (pmmmwh) [reply test/cache.ts]: Updated in types and tests.

File: source/utils/types.ts (hunk 19 above)
Comment (pmmmwh) [reply source/utils/types.ts]: Fixed accordingly 👍

File: source/utils/types.ts (hunk 18 above)
Comment (sindresorhus) [reply source/utils/types.ts]: Yes

File: source/create.ts [hunk 25]
```diff
@@ -24,36 +23,39 @@ export type HTTPAlias =
...
+export type ReturnStream = <T>(url: string | Merge<Options, {isStream?: true}>, options?: Merge<Options, {isStream?: true}>) => ProxyStream<T>;
+export type GotReturn<T = unknown> = CancelableRequest<T> | ProxyStream<T>;
 
 const getPromiseOrStream = (options: NormalizedOptions): GotReturn => options.isStream ? asStream(options) : asPromise(options);
 
-type OptionsOfDefaultResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType?: 'default'};
-type OptionsOfTextResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'text'};
-type OptionsOfJSONResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'json'};
-type OptionsOfBufferResponseBody = Options & {isStream?: false; resolveBodyOnly?: false; responseType: 'buffer'};
+const isGotInstance = (value: any): value is Got => (
```
Comment (sindresorhus) [inline source/create.ts]:
```suggestion
const isGotInstance = (value: unknown): value is Got => (
```

File: source/normalize-arguments.ts [hunk 26]
```diff
@@ -217,13 +224,13 @@ export const normalizeArguments = (url: URLOrOptions, options?: Options, default
 	if (is.urlInstance(url) || is.string(url)) {
 		options.url = url;
 
-		options = mergeOptions(defaults && defaults.options, options);
+		options = mergeOptions((defaults && defaults.options) ?? {}, options);
```
Comment (sindresorhus) [inline source/normalize-arguments.ts]:
```suggestion
		options = mergeOptions(defaults?.options ?? {}, options);
```

File: source/normalize-arguments.ts [hunk 27]
```diff
@@ -217,13 +224,13 @@ export const normalizeArguments = (url: URLOrOptions, options?: Options, default
...
 	} else {
 		if (Reflect.has(url, 'resolve')) {
 			throw new Error('The legacy `url.Url` is deprecated. Use `URL` instead.');
 		}
 
-		options = mergeOptions(defaults && defaults.options, url, options);
+		options = mergeOptions((defaults && defaults.options) ?? {}, url, options);
```
Comment (sindresorhus) [inline source/normalize-arguments.ts]:
```suggestion
		options = mergeOptions(defaults?.options ?? {}, url, options);
```

File: test/agent.ts [hunk 28]
```diff
@@ -29,7 +30,8 @@ const prepareServer = server => {
 	});
 };
 
-const createAgentSpy = Cls => {
+// TODO: Type this stricter if possible
+const createAgentSpy = <T extends any>(Cls: T) => {
```
Comment (sindresorhus) [inline test/agent.ts]:
```suggestion
const createAgentSpy = <T>(Cls: T) => {
```

And let's use a clearer name than `Cls`.

File: test/arguments.ts [hunk 29]
```diff
@@ -104,13 +107,13 @@ test('overrides `searchParams` from options', withServer, async (t, server, got)
...
-				get(key) {
+				get(key: string) {
 					t.is(key, `cacheable-request:GET:${server.url}/?test=wow`);
 				},
-				set(key) {
+				set(key: string) {
 					t.is(key, `cacheable-request:GET:${server.url}/?test=wow`);
 				}
-			}
+			} as unknown as StorageAdapter
```
Comment (sindresorhus) [inline test/arguments.ts]: Is this really needed? I think you just need to define the missing methods. The definition is:

```ts
interface Store<TValue> {
        get(key: string): TValue | Promise<TValue | undefined> | undefined;
        set(key: string, value: TValue, ttl?: number): any;
        delete(key: string): boolean | Promise<boolean>;
        clear(): void | Promise<void>;
    }
```

File: test/cookies.ts [hunk 30]
```diff
@@ -14,7 +15,7 @@ test('reads a cookie', withServer, async (t, server, got) => {
 
 	const cookieJar = new toughCookie.CookieJar();
 
-	await got({cookieJar});
+	await got({cookieJar} as unknown as OptionsOfDefaultResponseBody);
```
Comment (sindresorhus) [inline test/cookies.ts]: Why does this need force-casting?

File: test/create.ts [hunk 31]
```diff
@@ -251,14 +252,16 @@ test('async handlers', withServer, async (t, server, got) => {
 		handlers: [
 			async (options, next) => {
 				const result = await next(options);
+				// @ts-ignore Manual tests
 				result.modified = true;
 
 				return result;
 			}
-		]
+		] as HandlerFunction[]
```
Comment (sindresorhus) [inline test/create.ts]: Why does this need force-casting?

File: test/retry.ts [hunk 32]
```diff
@@ -68,7 +70,7 @@ test('retry function gets iteration count', withServer, async (t, server, got) =
 		retry: {
 			calculateDelay: ({attemptCount}) => {
 				t.true(is.number(attemptCount));
-				return attemptCount < 2;
+				return Number(attemptCount < 2);
```
Comment (sindresorhus) [inline test/retry.ts]: Why is this needed? `attemptCount` is not documented to ever be a different value or `undefined`.

File: test/types/create-test-server/index.d.ts [hunk 33]
```diff
@@ -0,0 +1,19 @@
+declare module 'create-test-server' {
+	import {Express} from 'express';
+
+	function createTestServer(options: any): Promise<createTestServer.TestServer>;
```
Comment (sindresorhus) [inline test/types/create-test-server/index.d.ts]:
```suggestion
	function createTestServer(options: unknown): Promise<createTestServer.TestServer>;
```

File: test/types/create-test-server/index.d.ts [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+
+	export = createTestServer;
+
+	namespace createTestServer {
+		export interface TestServer extends Express {
+			caCert: any;
```
Comment (sindresorhus) [inline test/types/create-test-server/index.d.ts]:
```suggestion
			caCert: unknown;
```

**Comment by sindresorhus** [comment]
There are still a lot of force-casting (`as`) in the tests that should ideally not be there.

File: test/agent.ts (hunk 28 above)
Comment (pmmmwh) [reply test/agent.ts]: Hmm. `unknown` doesn't seems to work here because it has to be something with a construct-able signature. Since `unknown` is not callable, it can never fulfill the type requirements?

File: test/cookies.ts (hunk 30 above)
Comment (pmmmwh) [reply test/cookies.ts]: I did this because I hit an issue where TypeScript is always using the first overload of the CookieJar's methods to match the use cases in our source code, which is actually using the third overload. Is there a way to make a proper comparison happen?

File: test/retry.ts (hunk 32 above)
Comment (pmmmwh) [reply test/retry.ts]: From reading the source it seems that the return value of `calculateDelay` would be consumed by `computedValue`, which expects a number. When we return `attemptCount < 2`, it is returning a `boolean` type, and to cast it to a number would need `Number()`, or use `attemptCount < 2 ? 1 : 0`

File: test/create.ts (hunk 31 above)
Comment (pmmmwh) [reply test/create.ts]: This can be resolved, it was my oversight.

File: test/arguments.ts (hunk 29 above)
Comment (pmmmwh) [reply test/arguments.ts]: This is resolved by implementing all methods.

File: test/retry.ts (hunk 32 above)
Comment (sindresorhus) [reply test/retry.ts]: `attemptCount < 2 ? 1 : 0` would be much clearer. Can you update https://github.com/sindresorhus/got#testing too?

File: test/create.ts (hunk 31 above)
Comment (sindresorhus) [reply test/create.ts]: Not clear to me what is complaining, but users should not need to force cast this.

File: test/cookies.ts (hunk 30 above)
Comment (sindresorhus) [reply test/cookies.ts]: Overloads are order-dependent, so maybe try changing the order? If this is in fact a TS issue, it would be good to try to find an issue on the TS issue tracker and add a code comment linking to it.

File: test/agent.ts (hunk 28 above)
Comment (sindresorhus) [reply test/agent.ts]: Ah, but it should at least be something more specific than `any` then.

**Comment by sindresorhus** [comment]
@pmmmwh I'm merging this now as tests are passing, to reduce potential merge conflicts for you and so that we can get this tested out in a beta version I plan to release now. Can you continue the remaining work in a new PR?

**Comment by sindresorhus** [comment]
Just want to say that you've done an amazing job so far 🙌✨

**Comment by pmmmwh** [comment]
> @pmmmwh I'm merging this now as tests are passing, to reduce potential merge conflicts for you and so that we can get this tested out in a beta version I plan to release now. Can you continue the remaining work in a new PR?

Sure! I will continue the work tomorrow probably - thanks for merging this!
